[FEATURE] Correct and improve vcf output (insertion calling 2/2)

eldariont opened the pull request:
This PR fixes #111 and adds the output of insertions into the VCF. Additionally, it correct several small errors in the VCF output and propagates the `min_var_length` parameter to the VCF output method.

Irallia reviewed (changes requested):
I'm happy to see, that we got an output file now! :tada:
Thanks for the corrections and their documentation in the code.
I have some comments for the `find_and_output_variants()` and also thougt, that we should start testing edge cases.

eldariont commented on include/variant_detection/variant_output.hpp:
```diff
@@ -4,26 +4,34 @@
  * \details Extracts genomic variants from given junction clusters.
  */
-void find_and_output_variants(std::vector<Cluster> const & clusters, std::ostream & out_stream);
+void find_and_output_variants(std::vector<Cluster> const & clusters,
+                              cmd_arguments const & args,
+                              std::ostream & out_stream);
```
Yeah, I thought about that for a while, too, and ended up passing the args because this method handles the actual variant calling and output. In the future, we might add a lot of arguments that need to be passed to the method, e.g.
- maximum variant length
- maximum tolerated inserted sequence size for deletions (currently hard-coded as 5)
- maximum tolerated distance between breakends for insertions (current hard-coded as 1)
- variant classes to output
- output of symbolic or sequences alleles
- ...

Irallia replied:
okay, thats a good argument. Thanks for adding some of these variables directly :)

eldariont commented on src/iGenVar.cpp:
```diff
@@ -9,17 +9,6 @@
 #include "variant_detection/variant_detection.hpp"                  // for detect_junctions_in_long_reads_sam_file()
 #include "variant_detection/variant_output.hpp"                     // for find_and_output_variants()
 
-struct cmd_arguments
-{
-    std::filesystem::path alignment_short_reads_file_path{""};
-    std::filesystem::path alignment_long_reads_file_path{""};
-    std::filesystem::path output_file_path{};
-    std::vector<detection_methods> methods{cigar_string, split_read, read_pairs, read_depth};   // default: all methods
-    clustering_methods clustering_method{simple_clustering};                                    // default: simple clustering method
-    refinement_methods refinement_method{no_refinement};                                        // default: no refinement
-    uint64_t min_var_length = 30;
-};
-
```
Regardless of whether we pass args to `find_and_output_variants()` or not: I wondered why the struct was defined in the cpp and not the hpp. Would it not be more common to define the fields of a struct in the hpp?

Irallia replied:
Yes you are right. The other structs we have also in the header files.

eldariont commented on src/variant_detection/variant_output.cpp:
```diff
@@ -16,41 +19,49 @@ void find_and_output_variants(std::vector<Cluster> const & clusters, std::ostrea
+                    //Deletion
+                    if (distance >= args.min_var_length &&
+                        distance < 1000000 &&
+                        insert_size < 5)
```
Good point, we should get rid of this hard-coded value and replace it with a CLI parameter. Yes, it means that we accept up to 4 insertions in a deletion, e.g.
```
ABCDEFGHI
AB-----HI  deletion of five bases
  RST      insertion of three bases instead
```

eldariont replied:
Added this in separate commit.

eldariont commented on src/variant_detection/variant_output.cpp:
```diff
@@ -16,41 +19,49 @@ void find_and_output_variants(std::vector<Cluster> const & clusters, std::ostrea
                         tmp.add_info("SVTYPE", "DEL");
-                        tmp.set_pos(mate1_pos);
-                        tmp.add_info("SVLEN", std::to_string(-distance));
+                        // Increment position by 1 because VCF is 1-based
+                        tmp.set_pos(mate1_pos + 1);
+                        tmp.add_info("SVLEN", std::to_string(-distance + 1));
+                        // Increment end by 1 because VCF is 1-based
+                        // Decrement end by 1 because deletion ends one base before mate2 begins
```
I totally agree 👍

eldariont replied:
Created an issue: #117

eldariont commented:
Hi @Irallia, I rebased this PR after the recent merges and incorporated your last three suggestions. Please let me know what you think about the other two suggestions where I added my reply 🙂
